Delete trashed orders after `EMPTY_TRASH_DAYS` as defined by WordPress (HPOS)

lsinger opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #41249.

Note / up for discussion: so far this doesn't yet do any batching and hooks into the WordPress-native `wp_scheduled_delete` hook. If we wanted to add batching we should probably go with an Action Scheduler job instead ([example from Blocks deleting draft orders here](https://github.com/woocommerce/woocommerce-blocks/blob/22793eb4ae5f8c2fbf19538e708d4b4489c725d2/src/Domain/Services/DraftOrders.php#L158-L186)). 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. the first commit contains a failing test -- review it for completeness and then check out this branch but walk back the HEAD ref so only the first commit is applied. 
2. run `pnpm --filter=woocommerce test:unit:env --filter=DataSynchronizerTests` to see the test fail
3. apply the reset of the commits from this branch
4. run `pnpm --filter=woocommerce test:unit:env --filter=DataSynchronizerTests` again and see the test succeed
5. alternatively trash an order under HPOS, change your server's time at least 30 days into the future, and ensure that the trashed order is being deleted by the respective job

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

lsinger committed:
Add a test that fails if trashed orders are never deleted under HPOS

lsinger committed:
Delete trashed orders after `EMPTY_TRASH_DAYS` as defined by WordPress

lsinger committed:
add the changelog file

lsinger committed:
appease the linter

jorgeatorres commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -966,6 +975,38 @@ private function delete_auto_draft_orders() {
+	 * @return void
+	 */
+	private function delete_trashed_orders() {
+		$delete_timestamp = $this->legacy_proxy->call_function( 'time' ) - ( DAY_IN_SECONDS * EMPTY_TRASH_DAYS );
```
I guess there's nothing technically wrong with having our own routine for deleting trashed orders regardless of datastore, but I think it'd be best if we check here that HPOS is authoritative before proceeding. We can let WP take care of legacy orders (which will propagate to HPOS when needed thanks to other code in the synchronizer).

There's `custom_orders_table_is_authoritative()` in this same class, which can be used for this purpose.

lsinger replied:
Good point! I like returning early whenever it makes sense. Addressed in [3c37df9](https://github.com/woocommerce/woocommerce/pull/41949/commits/3c37df96b57ecd631ec62b1426776d13bd2be4a8).

jorgeatorres commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -103,6 +110,7 @@ public function __construct() {
 		self::add_action( 'woocommerce_refund_created', array( $this, 'handle_updated_order' ), 100 );
 		self::add_action( 'woocommerce_update_order', array( $this, 'handle_updated_order' ), 100 );
 		self::add_action( 'wp_scheduled_auto_draft_delete', array( $this, 'delete_auto_draft_orders' ), 9 );
+		self::add_action( 'wp_scheduled_delete', array( $this, 'delete_trashed_orders' ) );
```
WP hooks onto `wp_scheduled_delete` with the default priority of 10 (see https://github.com/WordPress/wordpress-develop/blob/5eef2bbb2a476fc3f94f7d7c2d89ff01b9b74f82/src/wp-includes/default-filters.php#L432). I believe it'd be best for us to hook onto this a bit earlier (priority 9, for example) so that we fully delete things thanks to our sync code. Otherwise, we could end up in a situation where WP deletes the post record _before_ our own code deletes the order from the HPOS tables.
This shouldn't be a problem per se, but I'm thinking it might actually simplify things.

lsinger replied:
Yep, that does sound reasonable. ☺️ Addressed in [310ca5b](https://github.com/woocommerce/woocommerce/pull/41949/commits/310ca5ba502ea8234a6b38b1015494345d65328f).

jorgeatorres reviewed (changes requested):
Good job on this one @lsinger. Welcome to the wonderful world of HPOS. I left some comments, though mostly for us to discuss (nothing serious). Please let me know what you think!

lsinger committed:
return early if HPOS is not authorative

lsinger committed:
attempt to precede WordPress itself (which uses priority 10) to increase the probability of orders and their posts being handled solely by us

lsinger commented:
Thank you for the review and your insightful comments, @jorgeatorres -- I've rebased the PR to ensure we have the Blocks Merge in here, addressed your comments, and re-requested review. 👍

lsinger committed:
appease the linter

jorgeatorres reviewed (approved):
LGTM 🚢. Thank you @lsinger! 🙇

jorgeatorres commented:
By the way, given I don't yet have a time machine I tested this by enabling HPOS, trashing some orders and then modifying their modification date directly in the database (tables `wp_posts` and `wp_wc_orders`, as this has to be done to both the order and its backup post. I then ran `wp_scheduled_event` with the help of the WP Crontrol plugin.

In hindsight, maybe time travel would've been easier 😂
